src/Tree.hs

```diff
@@ -14,20 +22,40 @@ size (Node n _ _) = n
 node :: Tree a -> Tree a -> Tree a 
 node t1 t2 = Node (size t1 + size t2) t1 t2 
 
+-- Constructs a perfect tree from a list
+fromList :: [a] -> (Tree a, [a])
+fromList xs = buildTree (length xs) xs
+  where
+    buildTree 0 _  = error "Empty list in buildTree"
+    buildTree 1 (y:ys) = (Leaf y, ys)
+    buildTree n ys = 
+      let (leftSize, rightSize) = split n
+          (leftTree, rest) = buildTree leftSize ys
+          (rightTree, rest') = buildTree rightSize rest
+      in (node leftTree rightTree, rest')
+    split n = (n `div` 2, n `div` 2 + n `mod` 2)
+  
+-- >>> fromList "abcd"
+
 -- Flattens a tree into a list 
 toList :: Tree a -> [a]
-toList t = undefined 
+toList (Leaf x)       = [x]
+toList (Node _ xl xr) = toList xl ++ toList xr
 
 -- Fetches the k-th element in the tree 
 fetch :: Int -> Tree a -> a
-fetch k t = undefined 
+fetch k t = toList t !! k
 
 -- Updates the k-th element of a tree 
 update :: Show a => Int -> a -> Tree a -> Tree a 
-update k x t = undefined 
+update _ x (Leaf _)        = Leaf x
+update k x (Node sz xl xr)  = if size xl <= k 
+                               then Node sz xl (update (k - size xl) x xr)
+                               else Node sz (update k x xl) xr
 
 -- Checks that the tree is perfect, i.e. it contains exactly 2^n leaves, 
 -- and each node has two well-formed children of equal size. 
 -- Example of a perfect tree: Node 4 (Node 2 (Leaf 'c') (Leaf 'd')) (Node 2 (Leaf 'e') (Leaf 'f'))
 wellFormed :: Tree a -> Bool 
-wellFormed t = undefined 
\ No newline at end of file
+wellFormed (Leaf _)       = True
+wellFormed (Node _ xl xr) = size xl == size xr && wellFormed xl && wellFormed xr
```

Comment:
I'd also check that the size of the current node corresponds to the sizes of subtrees.